ios/browser/api/history/brave_history_api.mm

```diff
@@ -0,0 +1,229 @@
+#include "ui/base/page_transition_types.h"
+#include "url/gurl.h"
+
+#include "brave/ios/browser/api/history/brave_history_observer.h"
+#include "brave/ios/browser/api/history/history_service_listener_ios.h"
+
+#if !defined(__has_feature) || !__has_feature(objc_arc)
+#error "This file requires ARC support."
+#endif
+
+#pragma mark - IOSHistoryNode
+
+@interface IOSHistoryNode () {
+  std::u16string title_;
+  GURL gurl_;
+  base::Time date_added_;
+}
+@end
+
+@implementation IOSHistoryNode
+
+- (instancetype)initWithURL:(NSURL*)url
+                      title:(NSString* _Nullable)title
+                  dateAdded:(NSDate* _Nullable)dateAdded {
+  if ((self = [super init])) {
+    [self setUrl:url];
+
+    if (title) {
+      [self setTitle:title];
+    }
+
+    if (dateAdded) {
+      [self setDateAdded:dateAdded];
+    }
+  }
+
+  return self;
+}
+
+- (void)setUrl:(NSURL*)url {
+  gurl_ = net::GURLWithNSURL(url);
+}
+
+- (NSURL*)url {
+  return net::NSURLWithGURL(gurl_);
+}
+
+- (void)setTitle:(NSString*)title {
+  title_ = base::SysNSStringToUTF16(title);
+}
+
+- (NSString*)title {
+  return base::SysUTF16ToNSString(title_);
+}
+
+- (void)setDateAdded:(NSDate*)dateAdded {
+  date_added_ = base::Time::FromNSDate(dateAdded);
+}
+
+- (NSDate*)dateAdded {
+  return date_added_.ToNSDate();
+}
+@end
+
+#pragma mark - BraveHistoryAPI
+
+@interface BraveHistoryAPI () {
+  // History Service for adding and querying
+  history::HistoryService* history_service_;
+  // WebhistoryService for delete operations
+  history::WebHistoryService* web_history_service_;
+  // Tracker for history requests.
+  base::CancelableTaskTracker tracker_;
+}
+@property(nonatomic, strong) void (^query_completion)(NSArray<IOSHistoryNode*>*)
+    ;
+@end
+
+@implementation BraveHistoryAPI
+
+- (instancetype)initWithHistoryService:(history::HistoryService*)historyService
+                     webHistoryService:
+                         (history::WebHistoryService*)webHistoryService {
+  if ((self = [super init])) {
+    DCHECK_CURRENTLY_ON(web::WebThread::UI);
+    history_service_ = historyService;
+    web_history_service_ = webHistoryService;
+  }
+  return self;
+}
+
+- (void)dealloc {
+  history_service_ = nil;
+  web_history_service_ = nil;
+}
+
+- (id<HistoryServiceListener>)addObserver:(id<HistoryServiceObserver>)observer {
+  return [[HistoryServiceListenerImpl alloc] init:observer
+                                   historyService:history_service_];
+}
+
+- (void)removeObserver:(id<HistoryServiceListener>)observer {
+  [observer destroy];
+}
+
+- (bool)isBackendLoaded {
+  // Triggers backend to load if it hasn't already, and then returns whether
+  // it's finished loading.
+  return history_service_->BackendLoaded();
+}
+
+- (void)addHistory:(IOSHistoryNode*)history isURLTyped:(BOOL)isURLTyped {
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
+  DCHECK(history_service_->backend_loaded());
+
+  history::HistoryAddPageArgs args;
```

Comment:
nit: 

can we use this constructor instead of the default constructor - https://source.chromium.org/chromium/chromium/src/+/main:components/history/core/browser/history_types.h;l=357

My concern is that with updates to the chromium code, there might be other parameters like `floc_allowed` that will have default values that we'd like to override. Thoughts?

Reply:
Fix at https://github.com/brave/brave-core/pull/8869/commits/fc20bb01f2cc0b02fc3de72b993c7b3f457447d4